Step 1 , 2 : 지하철 노선도 구간 관리 기능 / 구간 추가 기능 심화

impati opened the pull request:
안녕하세요 리뷰어님!
저는 최준영이라고 합니다. 첫 번째 리뷰 요청이 많이 늦었지만 잘 부탁드립니다!

## 기능을 구현해온 과정
- step1 부터 step2 까지 최소한의 설계로 요구사항을 이해하고 빠르게 해결해나아갔습니다.
- 요구사항이 대체로 마무리가 되어 간 이후에 step2 에서 리팩토링을 진행하였습니다.  (일급 컬렉션 ,원시값 포장 , 1단계 indent 등등)
- 페어와 헤어지고 난 후 제가 읽었을 때 기준으로 가독성 리팩토링 해보았습니다.
- data.sql 파일을 삭제하고 fixture 를 두어 하나의 테스트 내에서 흐름을 보여주도록 리팩토링했습니다.
- 프론트 어플리케이션에 맞춰 변경된 DTO 는 별도로 정의하여 사용하였습니다.
## 질문 사항은 코드에 남기겠습니다!

grolarkim committed:
docs: README.md에 요구사항 작성

grolarkim committed:
test: SectionDaoTest 구간 생성 테스트 추가

grolarkim committed:
feat: SectionDao.insert 구현

grolarkim committed:
feat: LINE , STATION 더미 데이터 추가

grolarkim committed:
refactor: sectionDao.insert 리팩토링

grolarkim committed:
feat: Section.validate 테스트 추가 및 구현

grolarkim committed:
docs: Section 업데이트

grolarkim committed:
feat: SectionService.saveSection 테스트 및 구현

grolarkim committed:
feat : SectionController.createSection 테스트 및 구현

grolarkim committed:
fix: duplicatekeyexception 예외 임시 대응

grolarkim committed:
refactor: duplicatekeyexception 예외 대응

grolarkim committed:
feat: SectionDao.findLastSection 노선의 하행 종착역이 있는 구간 반환 쿼리 테스트 및 구현

grolarkim committed:
feat: SectionDao.existByLineIdAndStationId 노선과 역 아이디에 해당하는 구간이 존재 여부 반환 메소드 테스트 및 구현

grolarkim committed:
refactor: SectionDao.existByLineIdAndStationId 테스트 케이스 추가

grolarkim committed:
feat: SectionService.saveSection 검증 로직 테스트 및 구현

grolarkim committed:
refactor: SectionService.saveSection 검증 로직 메소드 분리

grolarkim committed:
docs: README.md 구간 등록 기능

grolarkim committed:
test: SectionDao.existByLineId 테스트 추가

grolarkim committed:
feat: SectionDao.findAllByLineId 테스트 및 구현

grolarkim committed:
feat: SectionDao.deleteById 테스트 및 구현

grolarkim committed:
feat: SectionService.deleteSection 테스트 및 구현

grolarkim committed:
refactor: SectionService.deleteSection 리팩토링

grolarkim committed:
feat: SectionController.deleteSection 구현 및 인수테스트

grolarkim committed:
docs: 구간 제거 기능 업데이트

grolarkim committed:
docs: README.md 스텝 1 완료

grolarkim committed:
docs: README.md 스텝 2 시작

grolarkim committed:
test: 구간 추가 제약사항 변경 - 정상 성공 테스트 추가

설명이 필요해서 ParameterizedTest 사용하지 않고 따로 구현

grolarkim committed:
test: 구간 추가 제약사항 변경 - 예외 발생 테스트 추가

grolarkim committed:
feat: SectionService.saveSection 구현

grolarkim committed:
test: SectionServiceTest findAllByLineId 정렬 테스트 추가

grolarkim committed:
feat: SectionService.findAllByLineId 정렬 구현

grolarkim committed:
refactor: SectionService 리팩토링

grolarkim committed:
refactor: LineService.saveLine에 첫 구간 생성 기능 추가 및 테스트

grolarkim committed:
refactor: Section 도메인을 간접 참조를 직접 참조로 변환

기존 Section 은 다른 domain 의 Id를 가지고 있는 상태에서
다른 domain 의 객체를 가지고 있는 상태로 변환 및 대응

grolarkim committed:
chore: 코드 리포맷팅

grolarkim committed:
feat: Sections 객체 생성 및 toStations 테스트 추가

grolarkim committed:
feat: Sections.toStations 정렬 기능 테스트 추가 및 구현

grolarkim committed:
feat: Sections.findLastSection 테스트 및 구현

grolarkim committed:
refactor: 일급 컬렉션 Sections을 사용하여 deleteSection 리팩토링

grolarkim committed:
refactor: Sections.sort 리팩토링

기존 완전 탐색을 활용하여 O(N**2)의 복잡도를 가지던 정렬을 Map을 활용하여 O(N)으로 개선하여 리팩토링

grolarkim committed:
feat: Section.cuttedSection 테스트 및 구현

grolarkim committed:
refactor: SectionService.saveSection Section의 cuttedSection을 활용해서 리팩토링

grolarkim committed:
docs: Step2 요구사항 업데이트

impati committed:
refactor: dto 패키지를 request , response 로 분리

impati committed:
refactor: 라인 생성 리팩토링 및 CreateLineRequest DTO 생성

impati committed:
fix: data.sql을 삭제하고 Fixture 도입으로 테스트에서 전체 흐름을 가진다.

impati committed:
refactor: 구간 추가 및 구간 삭제 리팩토링

impati committed:
refactor: Sections 리팩토링

impati committed:
refactor: 인수 테스트 리팩토링

impati committed:
refactor: final 키워드 추가 및 코드 컨벤션

impati committed:
test: 인수테스트 추가

Laterality commented on src/main/java/subway/application/LineService.java:
```diff
@@ -33,21 +53,23 @@ public List<Line> findLines() {
     }
 
-    public Line findLineById(Long id) {
+    public Line findLineById(final Long id) {
```
내부에서만 쓰이는 `public` 메서드네요.

impati replied:
전체적으로 외부에서 쓰이지 않는 메서드를 private 하게 변경하였습니다.

Laterality commented on src/main/java/subway/application/SectionService.java:
```diff
@@ -0,0 +1,119 @@
+        }
+    }
+
+    private void addIntersectionIfMatchInLine(final Section section, final Line line) {
```
서비스에서 처리하는 부분이 꽤 많아보이는데, 도메인에서 처리돼야 할 로직들이 서비스에서 처리되고 있는 것은 아닐까요?

impati replied:
Sections 에서 addSection 메서드를 만들어 해결했습니다.

Laterality commented on src/main/java/subway/dao/LineDao.java:
```diff
@@ -13,46 +13,47 @@
-        String sql = "update LINE set name = ?, color = ? where id = ?";
-        jdbcTemplate.update(sql, new Object[]{newLine.getName(), newLine.getColor(), newLine.getId()});
+        final String sql = "update LINE set name = ?, color = ? where id = ?";
+        jdbcTemplate.update(sql,
+                new Object[]{newLine.getName(), newLine.getColor(), newLine.getId()});
```
가변 인자를 사용할 수도 있네요.

```java
        jdbcTemplate.update(sql,
                newLine.getName(), newLine.getColor(), newLine.getId());
```

Laterality commented on src/main/java/subway/dao/SectionDao.java:
```diff
@@ -0,0 +1,128 @@
+                new Distance(section.getDistance()));
+    }
+
+    public boolean existAllOrNotingInLineBySection(final Line line, final Section section) {
```
DAO가 도메인에 관한 지식을 알고 있는 걸까요?

impati replied:
기능을 리팩토링하며 자연스럽게 삭제했습니다.
DAO 에 최대한 도메인에 관한 로직이 들어가지 않도록 주의하겠습니다!

Laterality commented on src/main/java/subway/dao/StationDao.java:
```diff
@@ -13,46 +13,45 @@
 
-    public Station findById(Long id) {
-        String sql = "select * from STATION where id = ?";
+    public Station findById(final Long id) {
```
DAO마다 단건 조회시 반환값이 없는 경우에 대한 처리가 다른데요.

null 반환, 예외, Optional.empty() 중 하나로 통일하면 어떨까요?

impati replied:
JDBC 단건 조회시 Optional 로 반환하여 사용하는 입장에서 적절한 처리를 해주기 위해서 Optional 로 반환하도록 결정했습니다.

Laterality commented on src/main/java/subway/domain/Distance.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    private void validate(final Long value) {
+        Assert.notNull(value, "거리는 null 일 수 없습니다.");
```
primitive 타입을 사용하면 null 체크가 필요 없지 않을까요?

impati replied:
좋은 방법이라고 생각합니다!

Laterality commented on src/main/java/subway/domain/Line.java:
```diff
@@ -1,24 +1,29 @@
     private String name;
     private String color;
 
     public Line() {
     }
```
필요한 생성자일까요?

impati replied:
다시한번 전체적으로 불필요한 메서드나 생성자에 대해 리팩토링을 수행했습니다!

Laterality commented on src/main/java/subway/dto/request/LineRequest.java:
```diff
@@ -1,6 +1,7 @@
-package subway.dto;
+package subway.dto.request;
 
 public class LineRequest {
```
`Line`에 대해 무엇을 요청하는 클래스일까요?

impati replied:
기존에 LineRequest 는 LineController 에서 라인을 생성하는 createLine 메서드에서 , updateLine 메서드에서 사용되었던 DTO 입니다. 이후에 프론트 코드에 따라 변경을 수행하다보니 라인을 생성할 때 요청 필드가 추가되어 CreateLineRequest 라는 별도의 DTO 를 만들어주었고 LineRequest 는 여전히 updateLine 에서만 사용되는 DTO 가 되었습니다.CreateLineRequest DTO 를 생성해주었으니 LineRequest 또한 UpdateLineRequest 로 변경해주었습니다.

Laterality commented on src/main/java/subway/error/SubWayExceptionHandler.java:
```diff
@@ -0,0 +1,28 @@
+    })
+    public ResponseEntity<String> handleArgumentExceptionException(IllegalArgumentException exception) {
+        return ResponseEntity.badRequest()
+                .contentType(MediaType.APPLICATION_JSON)
+                .body(exception.getMessage());
```
`Content-Type`과 응답 body의 형식이 일치할까요?

impati replied:
에러가 발생했을 때 JSON 형식으로 내려줄 수도 있겠지만 현재는 단순 메시지를 String 타입으로 내려주는 것도 괜찮다고 생각해서 MediaType.TEXT_PLAIN 으로 변경하였습니다.

Laterality commented on src/test/java/subway/fixture/LineFixture.java:
```diff
@@ -0,0 +1,17 @@
+import subway.dao.LineDao;
+import subway.domain.Line;
+
+public class LineFixture {
```
도메인의 픽스쳐라면 도메인 패키지에 위치하면 어떨까요?

Laterality reviewed (changes requested):
안녕하세요.

잘 구현해주셨습니다. 남겨드린 피드백 확인 부탁드립니다.

실패하는 테스트가 있는데, 함께 확인해주세요.

impati committed:
fix: 노선에 구간이 1개인 경우 삭제할 수 없는 예외를 IllegalArgumentException 로 Throw 하도록 변경함에 따라 테스트 변경

impati committed:
refactor: 불필요한 부분이나 놓쳤던 부분에 대한 리뷰 반영

- 불필요한 생성자 제거
- 외부에서 사용하지 않는 메서드 접근제어자 private 으로 변경
- ExceptionHandler 에서 에러 메시지를 단순 응답하므로 TEXT_PLAIN 으로 변경
- Fixture 패키지를 domain/Fixture 로 변경
- LineRequest -> UpdateLineRequest 로 변경
- 가능한 primitive 을 사용
- 뼈대 코드 리팩토링 (Dao 가변인자 , 트랜잭션)

impati committed:
refactor: 단건 조회시 Optional 반환으로 통일

impati committed:
refactor: 구간 추가 심화 기능을 Sections.addSection 에서 구현하는 것으로 리팩토링

impati committed:
refactor: 놓친 리팩토링

impati commented:
# 변경 사항
- 노선에 구간이 1개인 경우 삭제할 수 없는 예외를 IllegalArgumentException 로 Throw 하도록 변경함에 따라 테스트 변경
- 불필요한 생성자 제거
- 외부에서 사용하지 않는 메서드 접근제어자 private 으로 변경
- ExceptionHandler 에서 에러 메시지를 단순 응답하므로 TEXT_PLAIN 으로 변경
- Fixture 패키지를 domain/Fixture 로 변경 
- LineRequest -> UpdateLineRequest 로 변경
- 가능한 primitive 을 사용
- 뼈대 코드 리팩토링 (Dao 가변인자 , 트랜잭션)
- JDBC 단건 조회시 Optional 로 반환하도록 통일
- 구간 추가를 도메인인 Sections.addSection 을 호출하여 구현하고 변경을 감지해 데이터베이스에 반영하도록 변경

## 질문 
JDBC 단건 조회시 Optional 로 반환하여 사용하는 입장에서 적절한 처리를 해주기 위해서 Optional 로 반환하도록 결정했습니다.
하지만 이렇게 되면 XXXDao 에서 findById 와 같은 단건 조회 API 를 사용할 때 사용하는 입장에서 모두 OrElse() 혹은 orElseThrow() 를 사용하여 반복 처리해 줘야한다는 문제가 있습니다. 

예를 들어 여러 곳에서 아래 API 를 통해 역 정보를 직접 가져오고 대부분의 경우 예외를 던진다고 할 때 여러 곳에서 해당 코드가 반복되는 문제가 발생합니다.

```java
 stationDao.findById(stationId)
                .orElseThrow(() -> new IllegalStateException("역을 찾을 수 없습니다."));
```

이러한 문제를 해결하기 위해 제가 생각한 방법으로는 `EntityFinder` 와 같은 클래스를 두어 위와 같은 코드를 모아 놓고 Entity를 findById하는 곳에서  `EntityFinder` 를 통해 가져오면 이러한 문제를 해결할 수 있을 것이라고 생각했습니다. (실제로 도입하지는 않았습니다.) 실제로 이러한 문제를 해결하기 위한 좋은 방법이 있는지 궁금합니다.

```java
[EntityFinder.java]

...
...
public Staion findStationById(long stationId){
    stationDao.findById(stationId).orElseThrow(() -> new IllegalStateException("역을 찾을 수 없습니다."));
}

public Line findLineById(long lindId){
    lineDao.findById(lindId).orElseThrow(() -> new IllegalStateException("노선을 찾을 수 없습니다."));
}

public Section findSectionById(long sectionId){
    sectionDao.findById(sectionid).orElseThrow(() -> new IllegalStateException("구간을 찾을 수 없습니다."));
}

...
...
```

Laterality commented on src/main/java/subway/application/SectionService.java:
```diff
@@ -0,0 +1,86 @@
+        return SectionResponse.from(dirtyCheck(beforeSection, afterSection));
+    }
+
+    private Section dirtyCheck(List<Section> beforeSection, List<Section> afterSection) {
```
변경이 발생한 `Section`만 저장하기 위해 작성하신 것으로 보이는데, 이 부분은 서비스보다는 데이터를 저장하는 영역의 관심사가 아닐까요?

impati replied:
변경이 발생한 부분만 업데이트를 어디서 수행해야하는가를 결정하는 것이 제일 어려웠던 것 같습니다. 
변경이 발생한 부분을 Dao 에서 처리하도록 변경하였습니다.!

Laterality commented on src/main/java/subway/domain/Distance.java:
```diff
@@ -0,0 +1,49 @@
+
+    private static final Long MIN_DISTANCE = 1L;
+
+    private Long value;
```
필드와 getter에 래퍼 타입을 사용할 필요가 있을까요?

Laterality commented on src/main/java/subway/domain/Sections.java:
```diff
@@ -0,0 +1,187 @@
+    }
+
+    private void addIfMatchDownStation(final Section newSection, final List<Station> stations) {
+        if (stations.contains(newSection.getDownStation())) {
+            final Section originSection = sections.stream()
+                    .filter(section -> section.getDownStation().equals(newSection.getDownStation()))
+                    .findFirst()
+                    .orElseThrow(IllegalStateException::new);
+            final Line line = newSection.getLine();
+            final Station upstation = originSection.getUpStation();
+            final Station downStation = newSection.getUpStation();
+            final long distance = originSection.getDistance() - newSection.getDistance();
+            sections.remove(originSection);
+            sections.add(newSection);
+            sections.add(new Section(originSection.getId(), line, upstation, downStation, new Distance(distance)));
+        }
+        if (stations.contains(newSection.getUpStation())) {
+            sections.add(newSection);
+        }
```
`addIfMatchUpStation`와 비슷한 부분이 많은데요. 중복을 제거해서 메서드 크기를 줄여보면 어떨까요?

impati replied:
중복 제거를 위한 메서드 분리를 하고 싶었지만 그렇게 되면 중복 제거를 위한 메서드의 인자가 많아진다는 문제 때문에 고민했던 부분이었습니다. 
케이스에 따라서 적절히 메서드를 분리하고 최대한 중복을 해결할 수 있는 부분을 추출하며 리팩토링 했습니다.

Laterality commented on src/main/java/subway/dto/request/CreateLineRequest.java:
```diff
@@ -0,0 +1,41 @@
+package subway.dto.request;
+
+public class CreateLineRequest {
```
다른 요청 객체는 `[명사][동사]Reuqest` 형태인데 이 클래스는 순서가 다르네요. 네이밍 패턴을 맞추면 어떨까요?

Laterality commented on src/main/java/subway/error/SubWayExceptionHandler.java:
```diff
@@ -0,0 +1,28 @@
+    })
+    public ResponseEntity<String> handleArgumentExceptionException(IllegalArgumentException exception) {
+        return ResponseEntity.badRequest()
+                .contentType(MediaType.TEXT_PLAIN)
```
다른 컨트롤러에서는 Json으로 응답하는데, 오류인 경우에만 Plain text로 응답할 필요가 있을까요?

Laterality commented on src/main/java/subway/ui/SectionController.java:
```diff
@@ -0,0 +1,38 @@
+            @RequestBody final SectionRequest sectionRequest) {
+        final SectionResponse sectionResponse = sectionService.saveSection(lineId, sectionRequest);
+        return ResponseEntity.created(
+                        URI.create("/lines/" + lineId + "/sections/" + sectionResponse.getId()))
```
유효한 URI일까요?

Laterality commented on src/test/java/subway/integration/fixture/LineIntegrationFixture.java:
```diff
@@ -0,0 +1,25 @@
+
+public class LineIntegrationFixture {
+
+    public static Line createLine(final CreateLineRequest createLineRequest) {
```
픽스쳐보다는 테스트에 사용하는 헬퍼 메서드로 봐야 하지 않을까요?

impati replied:
Helper 와 Fixture 의 차이에 대해 잘 몰랐는데 이번 기회에 각각 어느 때 사용하는지 알아보겠습니다.!

Laterality reviewed (changes requested):
피드백 잘 반영해주셨습니다.

추가로 조금 더 개선해보면 좋을 만한 부분들이 있어 의견 남겨드렸어요. 

확인 부탁드립니다.

## Q&A

메서드 중복이 문제라고 생각하신다면 같은 클래스에 별도의 메서드로 분리하는 것에서부터 시작해보시면 어떨까요?

```java
private Line getLineById(long lineId) {
        final Line line = lineDao.findById(lineId)
                .orElseThrow(() -> new IllegalStateException("노선을 찾을 수 없습니다."));
}


```

impati committed:
refactor: 리뷰 반영 리팩토링

- id -> XXXId 로 변경 및 클래스 네이밍 통일
- distance 에서 래퍼클래스로 사용하는 부분을 모두 primitive 로 변경
- 잘못된 URI 를 수정 (SectionController.createSection)
- Integration Fixture -> Integration Helper 로 변경

impati committed:
feat: 에러응답을 statusCode , message 포함하도록 재정의

impati committed:
feat: SortedSections 를 도입하여 구간에 순서가 보장되는지를 표현

impati committed:
refactor: addSection 메서드 분리 리팩토링 및 DirtyChecking 을 Dao 에서 수행

impati committed:
refactor: 먼저 findById 를 메서드로 분리하는 것으로 Q&A 반영

impati commented:
# 변경 사항
- id -> XXXId 로 변경 및 클래스 네이밍 통일
- distance 에서 래퍼클래스로 사용하는 부분을 모두 primitive 로 변경
- 잘못된 URI 를 수정 (SectionController.createSection)
- Integration Fixture -> Integration Helper 로 변경
- 에러응답을 Json 형식으로 statusCode , message 포함하도록 변경
- 변경을 감지하고 반영하는 역할을 SectionDao 에서 하도록 변경
- Section.addSection 메서드 분리 리팩토링

# SortedSections vs Sections
현재 Sections.addSection 에서는 List < Section >  에 대해서 정렬된 상태가 필요없었습니다. 
뿐만 아니라  Sections 네이밍이 구간 정보가 정렬되어 있다는 정보를 외부에서 알기 어려웠다고 생각했습니다.
그래서 Sections 를 상속한 SortedSections 를 만들어 정렬된 상태가 필요한 곳에서 SortedSections 를 사용하도록 했습니다.

현재에는 SortedSections에만 하행 종점역을 삭제하는 API 가 있고 Sections 에 대한 삭제 API 가 없습니다. 이는 다음 단계에서 요구사항에 따라 추가적으로 변경해야할 부분이라고 생각했습니다.

Laterality commented on src/main/java/subway/application/LineService.java:
```diff
@@ -1,53 +1,81 @@
-    public LineResponse saveLine(LineRequest request) {
-        Line persistLine = lineDao.insert(new Line(request.getName(), request.getColor()));
+    @Transactional
+    public LineResponse saveLine(final LineCreateRequest request) {
```
새로 만드는 메서드와 갱신하는 메서드가 분리돼 있다면 생성하는 메서드의 이름은 save보다는 create가 직관적이지 않을까요?

Laterality commented on src/main/java/subway/application/StationService.java:
```diff
@@ -1,44 +1,67 @@
\ No newline at end of file
+
+    @Transactional(readOnly = true)
+    public List<Station> findStationByLineId(final Long lineId) {
```
1. 컬렉션을 반환한다면 메서드명은 복수형(`findStations~`)가 낫지 않을까요?
2. 습관적으로 래퍼 타입을 사용하고 계신 것 같은데, 굳이 사용할 필요가 있을까요?

impati replied:
id 의 경우
id 값을 primitive 값으로하면 없는 값에 대해 0으로 동작하여 잘못된 상황이 발생했을 때에도 오류없이 잘못된 동작하지만 오류를 발견하기 어렵다는 문제가 있다고 생각합니다. 그래서 잘못된 상황이 발생했을 때에 잘못된 값이 사용 되기전에 오류를 발생시켜 대응할 수 있도록 id 는 래퍼 타입으로 사용했습니다.

Laterality commented on src/main/java/subway/dao/SectionDao.java:
```diff
@@ -0,0 +1,105 @@
+        jdbcTemplate.update(sql, section.getUpStationId(), section.getDownStationId(), section.getDistance(), section.getId());
+    }
+
+    public Section dirtyChecking(Sections beforeSections, Sections afterSections) {
```
더티 체크는 DAO의 관심사인데 클라이언트가 알 필요가 있을까요?

Laterality commented on src/main/java/subway/domain/Distance.java:
```diff
@@ -0,0 +1,49 @@
+
+    private static final long MIN_DISTANCE = 1L;
+
+    private long value;
```
불변인 필드에는 `final`을 붙여 불변임을 명시적으로 나타내면 어떨까요?

Laterality commented on src/main/java/subway/domain/Line.java:
```diff
@@ -1,24 +1,26 @@
     }
 
-    public Line(Long id, String name, String color) {
+    public Line(final Long id, final String name, final String color) {
```
위의 생성자에서는 `id`에 null을 허용하는데 이 생성자에서는 허용하지 않는 이유가 무엇일까요?

impati replied:
id 가 없는 생성자는 아직 영속화되지 않았을 때를 도메인을 의미하고 
id 가 있는 생성자는 영속화되었을 때를 도메인을 의미합니다.

Laterality commented on src/main/java/subway/error/ErrorData.java:
```diff
@@ -0,0 +1,26 @@
+package subway.error;
+
+public class ErrorData {
```
`Data`, `Info`와 같은 단어가 이름에서 유의미하게 쓰이는 경우는 많지 않습니다. 클린코드 p.26을 참고해보세요.

Laterality commented on src/main/java/subway/error/SubWayExceptionHandler.java:
```diff
@@ -0,0 +1,43 @@
+            IllegalStateException.class
+    })
+    public ResponseEntity<ErrorData> handleIllegalStateExceptionException() {
+        return ResponseEntity.badRequest()
+                .contentType(MediaType.APPLICATION_JSON)
+                .body(ErrorData.of(INTERNAL_SERVER_ERROR.value(), "서버가 잘못 설계되었습니다. 문의 주십시오."));
```
1. 서버 측 문제로 요청에 실패했다면 BAD REQUEST가 적절한 상태 코드일까요?
2. 이 경우에는 문제가 생긴 지점을 파악해 수정이 필요할텐데요. 이 코드만으로 원인을 파악하기 용이할까요?

Laterality commented on src/main/java/subway/ui/SectionController.java:
```diff
@@ -0,0 +1,36 @@
+            @PathVariable final Long lineId,
+            @RequestBody final SectionRequest sectionRequest) {
+        final SectionResponse sectionResponse = sectionService.saveSection(lineId, sectionRequest);
+        return ResponseEntity.created(URI.create("/lines/" + lineId + "/sections")).body(sectionResponse);
```
201 상태 코드에 `Location` 헤더를 포함시켰다면 이 헤더의 URI가 리소스의 위치를 나타낸다는 의미일텐데요. 클라이언트는 이 URI를 사용해 리소스에 접근할 수 있나요?

impati replied:
구간 추가 성공시 리소스 위치에 대한 정보가 클라이언트 입장에서 필요 없다고 생각해서 구간 추가 시 성공했다는 상태 코드만 응답하도록 변경

Laterality commented on src/test/java/subway/domain/fixture/StationFixture.java:
```diff
@@ -0,0 +1,41 @@
+    private Station stationD;
+    private Station stationE;
+
+    public void init(StationDao stationDao) {
```
픽스쳐는 테스트용 데이터 준비라는 역할에만 충실하고, DAO에 이를 저장하는 일은 테스트에서 처리하면 어떨까요?

impati replied:
서비스 레이어에서 더티 체크 관심사를 제거하고 DAO 에서 하도록 변경했습니다.

Laterality reviewed (approved):
피드백 잘 반영해주셨습니다.

다음 단계 진행을 위해 이만 머지하도록 하겠습니다. 

추가로 개선해보면 좋을 것 같은 부분이 있어 코멘트 남겨드렸는데요. 다음 단계 진행하시면서 참고해주세요.
